GEODE-6660: Remove non-Linux OS statistics classes

alb3rtobr opened the pull request:
Thank you for submitting a contribution to Apache Geode.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [x] Is there a JIRA ticket associated with this PR? Is it referenced in the commit message?

- [x] Has your PR been rebased against the latest commit within the target branch (typically `develop`)?

- [x] Is your initial contribution a single, squashed commit?

- [x] Does `gradlew build` run cleanly?

- [x] Have you written or updated unit tests to verify your changes?

- [x] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?

### Note:
Please ensure that once the PR is submitted, check Concourse for build issues and
submit an update to your PR as soon as possible. If you need help, please send an
email to dev@geode.apache.org.

alb3rtobr commented:
Hi,
I have seen this pull request has not triggered the execution of the ci checks. As this is my first pull request Im wondering if I missed something...

pivotal-jbarrett commented:
Push an empty commit to retrigger.

metatype commented:
@dschneider-pivotal any concerns with merging these changes?

alb3rtobr commented:
Should I remove `init()` and `close()` too?

alb3rtobr commented:
I have realized that I could remove `isLinux()` too. The code in the static initializer block will throw an exception if the OS is not Linux, so if `isLinux()` is used, that means we are in a linux OS, so the method will always return `true`.

The static initializer block and the method:
```
static {
    String osName = System.getProperty("os.name", "unknown");
    if (!PureJavaMode.osStatsAreAvailable()) {
      throw new RuntimeException("HostStatHelper not allowed in pure java mode");
    } else if (osName.startsWith("Linux")) {
      osCode = LINUX_CODE;
    } else {
      throw new InternalGemFireException(
          String.format(
              "Unsupported OS %s. Only Linux(x86) OSs is supported.",
              osName));
    }
  }

public static boolean isLinux() {
    return osCode == LINUX_CODE;
  }
```

dschneider-pivotal commented:
Yes, remove all the native methods.

dschneider-pivotal commented:
At some point I think this class could be redesigned and it would just be LinuxStatHelper.
isLinux was being used by some code that I think now only needs to call PureJavaMode.osStatsAreAvailable(). I think PureJavaMode also needs to be cleaned up since that is the only mode we now support (geode never has the native library that used to make use not pure).
For this pr I would just focus on cleaning up HostStatHelper as much as you can and save the other cleanup for other prs.

alb3rtobr commented on geode-core/src/main/java/org/apache/geode/internal/statistics/HostStatHelper.java:
```diff
@@ -31,109 +30,55 @@
-
   static {
     String osName = System.getProperty("os.name", "unknown");
-    if (!PureJavaMode.osStatsAreAvailable()) {
```
I removed the `osStatsAreAvailable()` called because is redundant: if it returns `false`, that means that `osName.startsWith("Linux")` is `false`, which is what will be checked now.

alb3rtobr commented on geode-core/src/main/java/org/apache/geode/internal/statistics/HostStatHelper.java:
```diff
@@ -31,109 +30,55 @@
   }
 
   static int initOSStats() {
-    if (isLinux()) {
```
These `if (isLinux())` are not needed because that will be always true.

alb3rtobr commented on geode-core/src/main/java/org/apache/geode/internal/statistics/HostStatHelper.java:
```diff
@@ -154,10 +99,6 @@ public static void refresh(LocalStatisticsImpl statistics) {
   static Statistics newProcess(OsStatisticsFactory osStatisticsFactory, long pid, String name) {
     Statistics statistics;
-    if (osCode != LINUX_CODE) {
-      throw new InternalGemFireException(
```
I removed these `if (osCode != LINUX_CODE)` due to only linux is allowed.

alb3rtobr commented on geode-core/src/test/java/org/apache/geode/internal/statistics/OSVerifierTest.java:
```diff
@@ -0,0 +1,40 @@
+  }
+  
+  @Test
+  public void givenNonLinuxOs_thenOSVerifierObjectCanBeBuilt() {
```
The test title is wrong, it should be `givenNonLinuxOs_thenOSVerifierThrowsException`. I will fix it in next commit.

dschneider-pivotal commented on geode-core/src/test/java/org/apache/geode/internal/statistics/OSVerifierTest.java:
```diff
@@ -0,0 +1,55 @@
+  @Rule
+  public ExpectedException exceptionGrabber = ExpectedException.none();
+
+  @BeforeClass
```
I think a more standard way, in our test framework, to deal with tests that set system properties is:
  @Rule
  public RestoreSystemProperties restoreSystemProperties = new RestoreSystemProperties();

alb3rtobr commented:
> Are you still working on other commits or are you ready for a full review?

Yes, I have just fixed the wrong test name I detected and added the `@Rule` you suggested, thanks.

My only concern is if the HostStatHelper UT is enough for this PR. With the refactor I think the more important part of its behaviour is covered, anyway.

dschneider-pivotal commented on geode-core/src/main/java/org/apache/geode/internal/statistics/HostStatHelper.java:
```diff
@@ -17,154 +17,64 @@
-import org.apache.geode.internal.statistics.platform.WindowsSystemStats;
 
 /**
  * Provides native methods which fetch operating system statistics.
```
Add a better description of what this class is (it no longer has native methods).

pivotal-jbarrett commented on geode-core/src/main/java/org/apache/geode/internal/statistics/HostStatHelper.java:
```diff
@@ -17,154 +17,65 @@
+    new OSVerifier();
   }
 
   public static boolean isLinux() {
```
Is there still a need for a method that always returns true?

alb3rtobr replied:
I have changed the method in the last commit

alb3rtobr commented:
I have realized I have to change this code, or Geode will not be able to work on Windows.
With this code, an exception will be thrown whenever the `HostStatHelper` class is loaded, and the OS is not linux, which is not the expected behaviour.

kirklund commented on geode-core/src/main/java/org/apache/geode/internal/statistics/HostStatHelper.java:
```diff
@@ -32,28 +32,32 @@
 
   static {
-    new OSVerifier();
+    osVerifier = new OSVerifier();
```
In general, non-final statics are bad. Also, this line is simple enough that I would recommend you inline this static initializer and make osVerifier final:
```
private static final OSVerifier osVerifier = new OSVerifier();
```
Another option to consider (I'm not actually requesting this but I would help steer/review it if you choose to do this) is to change HostStatHelper from a utils class of static methods to an instance based object with non-static fields. This would make for better object design and improve testability of both the stat sampler as well as HostStatHelper.

For example, if HostStatHelper was renamed to OsStatisticsProvider, then GemFireStatSampler could change from:
```
Statistics stats =
    HostStatHelper.newProcess(getOsStatisticsFactory(), id, statName + "-proc");
processStats = HostStatHelper.newProcessStats(stats);
```
...to:
```
private final OsStatisticsProvider osStatisticsProvider = new OsStatisticsProvider();

Statistics stats = osStatisticsProvider.newProcess(getOsStatisticsFactory(), id, statName + "-proc");
processStats = osStatisticsProvider.newProcessStats(stats);
```
And then have the `OsStatisticsProvider` constructor read any System Properties to determine OS type.

Another good option is to have HostStatHelper and OsStatisticsFactory merge into a single class with no statics.

One major benefit to changing from statics: you can easily set a System Property in setUp for a test, then create the stat sampler or HostStatHelper/OsStatisticsFactory or any of these classes you intend to unit test. Since there's nothing that's `static final`, you can easily change the OS for many different tests in the same test class. When GemFireStatSampler executes the line `osStatisticsProvider = new OsStatisticsProvider();` the System Property would be read in the constructor of `OsStatisticsProvider`.

alb3rtobr replied:
> change HostStatHelper from a utils class of static methods to an instance based object with non-static fields

I like this idea, I also thought about it when I was dealing with the testing. And then the OsVerifier class will not be needed.

dschneider-pivotal commented on geode-core/src/main/java/org/apache/geode/management/internal/beans/MemberMBeanBridge.java:
```diff
@@ -430,14 +428,8 @@ public MemberMBeanBridge init() {
-      if (HostStatHelper.isSolaris()) {
-        systemStats = system.findStatisticsByType(SolarisSystemStats.getType());
-      } else if (HostStatHelper.isLinux()) {
+      if (HostStatHelper.isLinux()) {
```
This change should not have been needed. PureJavaMode.osStatsAreAvailable should now only return true if isLinux. Since os stat availability no longer has anything to do with PureJavaMode, I think you should move osStatsAreAvailable to the HostStatHelper. HostStatHelper is the class that should manage everything having to do with os stats. Also, have it be something like HostStatHelper.osStatsSupported instead of isLinux. In the future we might want to support os stats on platforms in addition to Linux so code outside of HostStatHelper should not be testing isLinux to decide if os stats are available.

alb3rtobr commented on geode-core/src/main/java/org/apache/geode/internal/statistics/OSVerifier.java:
```diff
@@ -18,15 +18,38 @@
 
 public class OSVerifier {
```
Thanks, I didnt know that class, Im going to take a look

alb3rtobr commented:
Last commit includes the following changes:
* Rename `HostStatHelper` to `OsStatisticsProvider` and transform it into an instance based object class.
* Change all calls to `PureJavaMode.osStatsAreAvailable` by `OsStatisticsProvider.osStatsSupported`.
* Remove `osStatsAreAvailable` from `PureJavaMode`.
* Now `OsStatisticsProvider` uses `SystemUtils`.
* Use `assertThat` instead of `assertTrue` in tests.

dschneider-pivotal commented on geode-core/src/main/java/org/apache/geode/internal/statistics/platform/LinuxProcessStats.java:
```diff
@@ -72,7 +72,7 @@ public static StatisticsType getType() {
                                                                           // package-protected
     if (stats instanceof LocalStatisticsImpl) {
-      HostStatHelper.refresh((LocalStatisticsImpl) stats);
+      OsStatisticsProvider.build().refresh((LocalStatisticsImpl) stats);
```
Please change this. The only code that calls this createProcessStats method is in OsStatisticsProvider. So move this first block of code (the refresh call) to that caller. That way we will not need to build a new instance of OsStatisticsProvider each time createProcessStats is called.

alb3rtobr replied:
+1, Im fixing this

dschneider-pivotal commented on geode-core/src/main/java/org/apache/geode/internal/statistics/OsStatisticsProvider.java:
```diff
@@ -27,47 +28,59 @@
-  private static final int SYSTEM_STAT_FLAG = 2;
-  private static final boolean isLinux;
+public class OsStatisticsProvider {
+  private final int PROCESS_STAT_FLAG = 1;
```
I think these two constants (PROCESS_STAT_FLAG and SYSTEM_STAT_FLAG) should still be static.

alb3rtobr replied:
What is the advantage of setting them as static? They are used only inside the OsStatisticsProvider class.

dschneider-pivotal replied:
All the instances of this class can share the same constant. You would normally only make something an instance variable if each instance needs its own copy. So the normal practice for constants is to make them static final.
